azurerm/internal/locks/lock.go

```diff
@@ -16,8 +20,11 @@ func ByName(name string, resourceType string) {
 }
 
 func MultipleByName(names *[]string, resourceType string) {
-	for _, name := range *names {
-		ByName(name, resourceType)
+	for i, name := range *names {
+		// at the end of every array item add its index. This way we guarantee that this item will be unique (no duplicates are possible)
+		uniqueValue := fmt.Sprintf("%s-arrIdx-%d", name, i)
```

Comment:
This has the same issue as the previous solution, if two resources call this with resource N, but they have different indexes we could deadlock. What we should do is just ensure there are no duplicates:
```
func SliceUniqString(elements []string) []sring {
	seen := make(map[int]bool, len(s))
	j := 0
	for _, v := range s {
		if _, ok := seen[v]; ok {
			continue
		}
		seen[v] = true
		elements[j] = v
		j++
	}
	return elements[:j]
}
```

Reply:
This also abstracts out the actual lock names, which could introduce hard to debug issues

Reply:
Right, I got your idea. I have a prepared similar solution. I was trying to avoid the deletion/remove of duplicated strings and that's why I proposed my previous solution but maybe you are right and this is the right way. I have committed the new changes

Reply:
@katbyte I know that you are busy working on v2.0.0 but just wanted to ask if I need to make some other changes or the last ones are OK? And do you think that this fix will be merged in the next v1.44? I'm asking because it is an essential bug for our team and currently blocking some of our tasks and I just want to have some more information.